File: 0701-insert-into-a-binary-search-tree/0701-insert-into-a-binary-search-tree.cpp
/**
 *Definition for a binary tree node.
 *struct TreeNode {
 *    int val;
 *    TreeNode * left;
 *    TreeNode * right;
 *    TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 *};
 */
class Solution
{
    public:
        TreeNode* insertIntoBST(TreeNode *root, int val)
        {
            
            //O(n) and space complexity O(1)
            TreeNode *node = new TreeNode(val);
            TreeNode *temp = root;
            
            if(root==NULL)
                return node;
            while (root != NULL)
            {
                if (val < root->val)
                {
                    if (root->left != NULL)
                        root = root->left;
                    else
                    {
                        root->left = node;
                        break;
                    }
                }
                else
                {
                    if (root->right != NULL)
                        root = root->right;
                    else
                    {
                        root->right = node;
                        break;
                    }
                }
            }
            return temp;
        }
};

File: 0701-insert-into-a-binary-search-tree/0701-insert-into-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode insertIntoBST(TreeNode root, int val) {
        
        TreeNode ptr = root;
        TreeNode node = new TreeNode(val);
        
        if(root==null){
            root = node;
            return root;
        }
        
        while(ptr!=null){
            
            if(ptr.val < val){
                if(ptr.right == null){
                    ptr.right = node;
                    return root;
                }
                else{
                    ptr = ptr.right;
                }
            }
            else {
                if(ptr.left == null){
                    ptr.left = node;
                    return root;
                }
                else{
                    ptr = ptr.left;
                }
            }
          
        }
        return root;
    }
}